crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,532 @@
+use kitsune2_api::{K2Error, K2Result};
+
+/// In-memory store for time slice hashes.
+///
+/// The inner store will look something like this:
+/// `[ Block_0_3, Single_3_ABC, Block_4_2, Single_6_DEF ]`
+///
+/// Consecutive sequences of empty hashes are compressed into a single block. A block
+/// contains at least 1 hash.
+///
+/// Gaps are not permitted, so the representation starts out as a single block that spans
+/// the entire range of possible slice ids. As hashes are inserted, the block is split around
+/// the inserted hash.
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone))]
+pub(super) struct TimeSliceHashStore {
+    pub(super) highest_stored_id: Option<u64>,
+    inner: Vec<SliceHash>,
+}
+
+impl Default for TimeSliceHashStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum SliceHash {
+    Single { id: u64, hash: bytes::Bytes },
+    Block { id: u64, count: u64 },
+}
+
+#[cfg(test)]
+impl SliceHash {
+    pub(crate) fn id(&self) -> u64 {
+        match self {
+            SliceHash::Single { id, .. } => *id,
+            SliceHash::Block { id, .. } => *id,
+        }
+    }
+
+    pub(crate) fn end(&self) -> u64 {
+        self.id()
+            + match self {
+                SliceHash::Single { .. } => 0,
+                SliceHash::Block { count, .. } => *count,
+            }
+    }
+}
+
+impl TimeSliceHashStore {
+    pub(super) fn new() -> Self {
+        Self {
+            highest_stored_id: None,
+            inner: vec![SliceHash::Block {
+                id: 0,
+                count: u64::MAX,
+            }],
+        }
+    }
+
+    /// Insert a hash at the given slice id.
+    pub(super) fn insert(
+        &mut self,
+        slice_id: u64,
+        hash: bytes::Bytes,
+    ) -> K2Result<()> {
+        if !hash.is_empty() {
+            match &mut self.highest_stored_id {
+                Some(id) if slice_id > *id => {
+                    *id = slice_id;
+                }
+                None => {
+                    self.highest_stored_id = Some(slice_id);
+                }
+                _ => {}
+            }
+        }
+
+        let mut handled = false;
+        for (i, slice_hash) in self.inner.iter().cloned().enumerate() {
```

Comment:
Not sure if it would be clearer or more efficient to use [position](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.position) rather than iterating from the start, or its effectively doing the same thing and you prefer to spell it out

Reply:
This was a mutable loop at some point and that became something the compiler couldn't reason about. I guess the slight downside with position is needing do do multiple match blocks, but it would simplify the code regardless

Reply:
Done: [89d71f3b](https://github.com/holochain/kitsune2/pull/37/commits/89d71f3b1ff4be3f8e812a0d3ff33f515b7709cd)

Reply:
I went with `rfind`, because most of the time I expect it to be newer time slices being updated, so it probably makes sense to search backwards.

Reply:
Also gone :)